lib/llvmopencl/SafeDivision.cc

```diff
@@ -0,0 +1,184 @@
+// LLVM function pass that ensures integer division by zero (or INT_MIN/-1)
+// does not result in signal sent to the process.
+//
+// Copyright (c) 2024 Michal Babej / Intel Finland Oy
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to
+// deal in the Software without restriction, including without limitation the
+// rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+// sell copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+
+#include "CompilerWarnings.h"
+IGNORE_COMPILER_WARNING("-Wmaybe-uninitialized")
+#include <llvm/ADT/Twine.h>
+POP_COMPILER_DIAGS
+IGNORE_COMPILER_WARNING("-Wunused-parameter")
+#include <llvm/ADT/SmallPtrSet.h>
+#include <llvm/IR/IRBuilder.h>
+#include <llvm/IR/InstrTypes.h>
+#include <llvm/IR/Module.h>
+#include <llvm/Pass.h>
+#include <llvm/Transforms/Utils/ValueMapper.h>
+
+#include "LLVMUtils.h"
+#include "SafeDivision.h"
+#include "WorkitemHandlerChooser.h"
+POP_COMPILER_DIAGS
+
+#include <iostream>
+#include <set>
+#include <string>
+
+#include "pocl_llvm_api.h"
+
+// #define DEBUG_SAFE_DIVISION
+
+#define PASS_NAME "safe-division"
+#define PASS_CLASS pocl::SafeDivision
+#define PASS_DESC "Handle integer division&remainder UB in kernel code"
```

Comment:
Or: "Change division and remainder instructions to non-trapping ones"

Reply:
I think that would be a misleading description. We're not changing instructions; also it implies that UB results in trap - which is only true on some architectures. However this pass is architecture-independent. So IMO the current description is more correct - it turns UB corner cases into defined behavior. It might even make sense to always enable it - it would lead to identical results of div/rem math across all architectures.